File: forge/src/main/java/net/impleri/itemskills/integrations/curios/CuriosSkills.java
package net.impleri.itemskills.integrations.curios;

import net.impleri.itemskills.InventoryHelper;
import net.impleri.itemskills.ItemHelper;
import net.minecraft.world.entity.player.Player;
import net.minecraftforge.common.util.NonNullConsumer;
import net.minecraftforge.items.IItemHandlerModifiable;
import top.theillusivec4.curios.api.CuriosApi;

public class CuriosSkills {
    public static void onPlayerTick(Player player) {
        CuriosApi.getCuriosHelper().getEquippedCurios(player).ifPresent(handleCurios(player));
    }

    private static NonNullConsumer<IItemHandlerModifiable> handleCurios(Player player) {
        return (IItemHandlerModifiable curios) -> {
            var slotCount = curios.getSlots();

            if (slotCount > 0) {
                for (int idx = 0; idx <= slotCount; idx++) {
                    var slot = curios.getStackInSlot(idx);

                    if (!ItemHelper.isWearable(player, slot.getItem(), null)) {
                        var badItem = curios.extractItem(idx, slot.getCount(), false);

                        InventoryHelper.moveItemIntoInventory(player, badItem);
                    }
                }
            }
        };
    }
}

File: forge/src/main/kotlin/net/impleri/itemskills/integrations/curios/CuriosSkills.kt
package net.impleri.itemskills.integrations.curios

import net.impleri.itemskills.api.ItemRestriction
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraftforge.items.IItemHandlerModifiable
import top.theillusivec4.curios.api.CuriosApi

object CuriosSkills {
  fun onPlayerTick(player: Player) {
    CuriosApi.getCuriosHelper().getEquippedCurios(player).ifPresent(handleCurios(player))
  }

  private fun handleCurios(player: Player): (IItemHandlerModifiable) -> Unit {
    return {
      val slotCount = it.slots
      if (slotCount > 0) {
        for (idx in 0..slotCount) {
          val slot: ItemStack = it.getStackInSlot(idx)
          if (!ItemRestriction.canEquip(player, slot, null)) {
            val badItem = it.extractItem(idx, slot.count, false)
            ItemRestriction.moveItemIntoInventory(player, badItem)
          }
        }
      }
    }
  }
}
